### PR TITLE
Refactors the code for gas optimisation.

### PR DESCRIPTION
**Pull request by afmfe-iul**
# Results
(all values, except "runs" column, expressed in gas)

#### Deployments
```
                            initial          final         saving
Fast                   |   2_341_663   ·   2_341_663   ·    0,00%
FastAccessFacet        |   1_624_627   ·   1_202_978   ·   25,95%
FastFrontendFacet      |   1_266_962   ·     968_394   ·   23,57%
FastHistoryFacet       |   1_880_857   ·   1_298_675   ·   30,95%
FastTokenFacet         |   4_580_376   ·   3_508_391   ·   23,40%
FastTopFacet           |     442_135   ·     312_393   ·   29,34%
IssuerAccessFacet      |     958_949   ·     631_827   ·   34,11%
IssuerFrontendFacet    |     755_673   ·     556_301   ·   26,38%
IssuerTopFacet         |     945_724   ·     652_858   ·   30,97%
MarketplaceAccessFacet |   1_713_025   ·   1_223_002   ·   28,61%
MarketplaceTopFacet    |     118_207   ·      97_073   ·   17,88%
```

#### Function calls
```
                           runs      initial        final       saving
addGovernor            |    32   ·   108_956   ·   106_684   ·   2,09%
addMember              |    82   ·    93_244   ·    91_519   ·   1,85%
addTransferCredits     |    60   ·    72_357   ·    70_592   ·   2,44%
approve                |    64   ·   161_419   ·   158_650   ·   1,72%
beforeRemovingMember   |     5   ·    87_583   ·    84_300   ·   3,75%
burn                   |     5   ·   163_632   ·   160_652   ·   1,82%
burnt                  |     1   ·   139_005   ·   138_577   ·   0,31%
diamondCut             |    43   ·   213_374   ·   213_374   ·   0,00%
disapprove             |     6   ·    58_487   ·    56_567   ·   3,28%
drainTransferCredits   |     5   ·    51_796   ·    50_327   ·   2,84%
emitDetailsChanged     |     2   ·    40_033   ·    39_244   ·   1,97%
mint                   |    66   ·   181_042   ·   177_734   ·   1,83%
minted                 |    13   ·   108_579   ·   108_140   ·   0,40%
removeGovernor         |     5   ·    58_987   ·    56_925   ·   3,50%
removeMember           |    13   ·    57_812   ·    55_719   ·   3,62%
setIsSemiPublic        |     4   ·    60_364   ·    58_373   ·   3,30%
transfer               |     1   ·   275_438   ·   267_326   ·   2,95%
transfered             |    33   ·   232_399   ·   231_427   ·   0,42%
transferFrom           |    88   ·   322_647   ·   311_963   ·   3,31%
transferFromWithRef    |    14   ·   287_540   ·   276_707   ·   3,77%
transferWithRef        |     1   ·   276_270   ·   268_245   ·   2,90%
initialize             |     5   ·   199_899   ·   197_918   ·   0,99%
registerFast           |    27   ·   112_201   ·   110_685   ·   1,35%
activateMember         |     3   ·    50_579   ·    48_337   ·   4,43%
deactivateMember       |    11   ·    92_136   ·    90_107   ·   2,20%
memberAddedToFast      |     5   ·    79_745   ·    78_785   ·   1,20%
memberRemovedFromFast  |     1   ·    37_878   ·    36_864   ·   2,68%
```

# Types and comments about optmisations done

### Enabling the Solidity optimizer
Not much to say here, it's advisable to use the optimizer even with the Diamond pattern, but there's room for tweaking the actual value: a smaller one will lead to smaller bytecode (cheaper contract deployments) and a bigger one will lead to cheaper function calls, right now I've used a really big number for cheaper calls, since we didn't exceed the max bytecode size on any deployed contract.

### Use of custom errors with revert
Solidity 0.8.4 added custom errors and after reading about it, which cost less gas and open the door for some interesting stuff like having arguments passed to the error or having better testing (soon with TypeChain!). Here are some links that explain it way better than I can:
- [Custom errors in Solidity](https://blog.soliditylang.org/2021/04/21/custom-errors/)
- [Feature Request on TypeChain for Custom errors](https://github.com/dethcrypto/TypeChain/issues/667) (potentially less brittle tests and error types on the frontend!)

### Unchecked Math where applicable
Mainly for looping and when updating balances after a check was already done. Lots of tips taken from [this post](https://betterprogramming.pub/the-ultimate-100-point-checklist-before-sending-your-smart-contract-for-audit-af9a5b5d95d0)
Maybe [this one](https://github.com/ethereum/solidity/issues/11721#issuecomment-890917517) is a better way of doing it, for readability.

### Better caching of variables
Stopped storing some values on variables that weren't used more than once, and started storing array lengths on variables before looping.

# Extra notes
1. Broke lots of tests because the previous revert messages don't match the Custom errors. Didn't fix them because I'm not sure if we want to go for this change.
2. Left `LibDiamond.sol` un-touched but there are 14 `require` statements there that we could update (the contract was written prior to 0.8.4)
3. I'm not sure if we looked at the different Diamond implementations and picked the best suited for us ([Diamond Implementations](https://github.com/mudgen/diamond), or even if with the tooling we use if we can choose one. As an example [Dark Forest](https://github.com/darkforest-eth/eth/tree/master/contracts) went for the diamond-2 implementation and have a note on their readme saying:
>Note: The implementations of diamondCut and loupe functions can vary based on complexity and gas cost choices. Dark Forest has chosen the ["diamond-2"](https://github.com/mudgen/diamond#diamond-implementations) implementation because it allows cheaper gas cost of the diamondCut function, which is useful for Lobbies. This results in a higher cost loupe function call.

### CODE REVIEW

File: contracts/fast/FastTokenFacet.sol [hunk 1]
```diff
@@ -50,11 +49,17 @@ contract FastTokenFacet is AFastFacet, IERC20, IERC1404 {
...
+    if(FastTopFacet(address(this)).hasFixedSupply()){
+      revert ICustomErrors.RequiresContinuousSupply();
+    }
+    if(balanceOf(address(0)) < amount){
+      revert ICustomErrors.InsufficientFunds();
+    }
 
     // Remove the minted amount from the zero address.
-    s.balances[address(0)] -= amount;
+    unchecked {
```
[the remark this answers is not in the document]
Comment (afmfe-iul) [reply contracts/fast/FastTokenFacet.sol:60]: Good point, that's even better probably, I'll double check!

File: contracts/fast/FastTokenFacet.sol [hunk 2]
```diff
@@ -277,48 +283,51 @@ contract FastTokenFacet is AFastFacet, IERC20, IERC1404 {
...
-      LibConstants.INSUFFICIENT_FUNDS
-    );
-    require(
-      p.amount > 0,
-      LibConstants.UNSUPPORTED_OPERATION
-    );
+    if(s.balances[p.from] < p.amount){
+      revert ICustomErrors.InsufficientFunds();
+    }
+    if(p.amount <= 0){
```
[the remark this answers is not in the document]
Comment (afmfe-iul) [reply contracts/fast/FastTokenFacet.sol:289]: Well spotted!

File: contracts/interfaces/ICustomErrors.sol [hunk 3]
```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.4;
+interface ICustomErrors {
+  error AlreadyInitialized();
```
[the remark this answers is not in the document]
Comment (afmfe-iul) [reply contracts/interfaces/ICustomErrors.sol:4]: Yep! For a while we probably need to hard code them, but as soon as TypeChain includes custom errors on the generated types, this is going to be smooooth! :)

File: contracts/fast/FastFrontendFacet.sol [hunk 4]
```diff
@@ -81,46 +81,45 @@ contract FastFrontendFacet is AFastFacet {
...
   function paginateDetailedMembers(uint256 index, uint256 perPage)
       external view returns(MemberDetails[] memory, uint256) {
-    LibFastAccess.Data storage accessStorage = LibFastAccess.data();
     (address[] memory members, uint256 nextCursor) =
-      LibPaginate.addresses(accessStorage.memberSet.values, index, perPage);
+      LibPaginate.addresses(LibFastAccess.data().memberSet.values, index, perPage);
     MemberDetails[] memory values = new MemberDetails[](members.length);
-    for (uint256 i = 0; i < members.length; ++i) {
+    uint256 length = members.length;
+    for (uint256 i = 0; i < length;) {
```
Comment (jamesduncombe) [inline contracts/fast/FastFrontendFacet.sol:107]: `for (uint256 i; i < length;) {`  ? using default `0` for uint256... would this be cheaper still?

Comment (hickscorp) [reply contracts/fast/FastFrontendFacet.sol:107]: Well if it's a gas optimisation, @afmfe-iul can have a look.
Just so you know - I'm against some of them.
If we're really after gas optimisations, we shouldn't have the `FastFacetHistory` - which is what consumes half of the gas during transfers - and could be removed if our users were tech-savy ;)

The `unchecked` keyword is **to be avoided**, unless we have the money to hire super-soldiers in solidity.
I only use `unchecked` in very specific circumstances. For example consider the following:

```solidity
require(balance >= amount, "Not enough funds or somethn'");
unchecked {
  balance -= amount;
}
```

So far so good. But as the complexity increases, and as multiple developers work and solve conflicts on the file, you could end up with:

```solidity
require(balance >= amount, "Not enough funds or somethin");
balance -= fee;
unchecked {
  balance -= amount;
}
```

Or even, why not:

```solidity
balance -= fee;
unchecked {
  balance -= amount;
}
```

As much as I like gas optimizations, the compiler isn't smart enough to fence the developer against these problems. So I'd rather have our users spend a bit more gas on the first snippet, or on:

```solidity
balance -= fee;
```

**Pull request merged**